go/plugins/firebase/firebase.go

```diff
@@ -16,31 +16,90 @@ package firebase
 
 import (
 	"context"
+	"fmt"
+	"log"
 	"sync"
 
 	firebase "firebase.google.com/go/v4"
 	"firebase.google.com/go/v4/auth"
+	"github.com/firebase/genkit/go/ai"
 )
 
+var state struct {
+	mu         sync.Mutex     // Ensures thread-safe access to state
+	initted    bool           // Tracks if the plugin has been initialized
+	app        *firebase.App  // Holds the Firebase app instance
+	retrievers []ai.Retriever // Holds the list of initialized retrievers
+}
+
+// FirebaseApp is an interface to represent the Firebase App object
 type FirebaseApp interface {
 	Auth(ctx context.Context) (*auth.Client, error)
 }
 
-var (
-	app   *firebase.App
-	mutex sync.Mutex
-)
+// FirebasePluginConfig is the configuration for the Firebase plugin.
+type FirebasePluginConfig struct {
+	App        *firebase.App      // Pre-initialized Firebase app
+	Retrievers []RetrieverOptions // Array of retriever options
+}
+
+// Init initializes the plugin with the provided configuration.
+func Init(ctx context.Context, cfg *FirebasePluginConfig) error {
+	state.mu.Lock()
+	defer state.mu.Unlock()
 
-// app returns a cached Firebase app.
-func App(ctx context.Context) (FirebaseApp, error) {
-	mutex.Lock()
-	defer mutex.Unlock()
-	if app == nil {
-		newApp, err := firebase.NewApp(ctx, nil)
+	if state.initted {
+		log.Println("firebase.Init: plugin already initialized, returning without reinitializing")
```

Comment:
We discussed this before, but I don't remember the outcome of the discussion -- are we sure we want to no-op when it's possible there is a different config being passed in?

Reply:
> I wonder if this is the right behavior or if it should error on subsequent calls. Generally this no-op behavior is only deterministic if you are not passing a configuration that you are expecting to take effect. Here, a no-op means you don't know if the initted config matches what you're expecting.

yup you're right, sorry it got lost across versions of this PR, will look to change